### PR TITLE
[AzCopyV10] Make piping more efficient with concurrency

### PR DESCRIPTION
**Pull request by siminsavani-msft**
- Made piping efficient and configurable for download & upload.
- For **Upload**: extracted _concurrencyValue_ from environment variable _AZCOPY_CONCURRENCY_VALUE_ and replaced pipingUploadParallelism with this value. 
- For **Download**: extracted _concurrencyValue_ from environment variable _AZCOPY_CONCURRENCY_VALUE_ and utilized the high-level _DoBatchTransfer_ function from Azure Go SDK (this code imitates logic in _DownloadBlobToFile_ function).

### CODE REVIEW

**Comment by mohsha-msft** [comment]
Hey @siminsavani-msft 

Did you try transferring large file through pipe upload/download?
If yes what were the perf improvements from before?

**Comment by siminsavani-msft** [comment]
> Hey @siminsavani-msft
> 
> Did you try transferring large file through pipe upload/download?
> If yes what were the perf improvements from before?

Here are the results from perf testing:

_Format: (size, concurrency value, new/old code, upload/download): minute:seconds_

Test 1 (200 MB file, CV = 1, new, upload ): 3:07

Test 2 (200 MB file, CV=3, new, upload): 1:39

Test 3 (200 MB file, CV=1, new, download): 3:20 

Test 4 (200 MB file, CV=3, new, download): 1:01

Test 5 (200 MB file, CV=0, old, upload): 3:21

Test 6 (200 MB file, CV=0, old, download): 1:02

**Review by mohsha-msft — APPROVED** [review]
LGTM. Just try once by removing `AZCOPY_CONCURRENCY_VALUE` from environment variables to see the default behavior.

File: cmd/copy.go [hunk 1]
```diff
@@ -1136,18 +1138,50 @@ func (cca *cookedCopyCmdArgs) processRedirectionDownload(blobResource common.Res
...
+			ChunkSize:     cca.blockSize,
+			Parallelism:   uint16(concurrencyValue),
+			Operation: func(chunkStart int64, count int64, ctx context.Context) error {
+				dr, err := blobURL.Download(ctx, chunkStart, count, azblob.BlobAccessConditions{}, false, clientProvidedKey)
+				if err != nil {
+					return err
+				}
+				body := dr.Body(azblob.RetryReaderOptions{MaxRetryRequests: ste.MaxRetryPerDownloadBody})
+				// step 4: pipe everything into Stdout
+				_, err = io.Copy(os.Stdout, body)
```
Comment (zezha-msft) [inline cmd/copy.go:1173]: Wait, I don't think this is going to work. If each chunk is just writing to the Stdout whenever they are done, doesn't that corrupt the file content?

**Review by zezha-msft — CHANGES_REQUESTED** [review]

File: cmd/copy.go (hunk 1 above)
Comment (zezha-msft) [reply cmd/copy.go:1173]: I think we need a synchronized writer, to write all the chunks out serially (they can be downloaded in parallel).

**Review by zezha-msft — CHANGES_REQUESTED** [review]

**Comment by mohsha-msft** [comment]
Let's merge this in next release may be.

**Comment by siminsavani-msft** [comment]
Here are the new perf results: https://microsoft-my.sharepoint.com/:x:/p/siminsavani/EQPBpJOeHb5AvCYHHzJIxcUBn4-t4Qt0UAIC_4BMq1nOhw?e=0fUC62

File: cmd/copy.go [hunk 2]
```diff
@@ -1165,18 +1166,84 @@ func (cca *cookedCopyCmdArgs) processRedirectionDownload(blobResource common.Res
...
+			Parallelism:  uint16(concurrencyValue),
+			Operation: func(chunkStart int64, count int64, ctx context.Context) error {
+				dr, err := blobURL.Download(ctx, chunkStart, count, azblob.BlobAccessConditions{}, false, clientProvidedKey)
+				if err != nil {
+					return err
+				}
+				body := dr.Body(azblob.RetryReaderOptions{MaxRetryRequests: ste.MaxRetryPerDownloadBody})
+
+				// step 4: pipe everything into Stdout
+				err = dstWriter.EnqueueChunk(ctx, common.ChunkID{}, cca.blockSize, os.Stdout, true)
```
Comment (adreed-msft) [inline cmd/copy.go:1227]: Chunk IDs are not unused in the chunked file writer, and this kinda ignores the whole point of using the chunked file writer.

Leaving this empty is incorrect and will cause data corruption. Please look at the ChunkID struct, and properly fill it out as much as is feasible here.

File: cmd/copy.go [hunk 3]
```diff
@@ -1165,18 +1166,84 @@ func (cca *cookedCopyCmdArgs) processRedirectionDownload(blobResource common.Res
...
+				if err != nil {
+					return err
+				}
+				_, err = dstWriter.Flush(ctx)
```
Comment (adreed-msft) [inline cmd/copy.go:1231]: Flush should not be done until after the batch transfer has completed.

**Review by adreed-msft — CHANGES_REQUESTED** [review]

**Review by adreed-msft — APPROVED** [review]
Otherwise looks good.

**Comment by zezha-msft** [comment]
Need perf test results for merging.

**Pull request closed without merging**